client/landing/stepper/hooks/use-checkout.ts

```diff
@@ -36,13 +38,15 @@ const useCheckout = () => {
 		setSignupCompleteFlowName( flowName );
 		setSignupCompleteStepName( stepName );
 
-		// If the plan is not provided, we might have added plan to the cart so we just go to the checkout page directly
-		if ( ! plan ) {
+		// If the plan not extra products are provided, we might have added plan to the cart so we just go to the checkout page directly
+		if ( ! plan && extraProducts.length === 0 ) {
 			// The theme upsell link does not work with siteId and requires a siteSlug.
 			// See https://github.com/Automattic/wp-calypso/pull/64899
 			window.location.href = `/checkout/${ encodeURIComponent( siteSlug ) }?${ params }`;
+		} else if ( ! plan ) {
+			openCheckoutModal( extraProducts, { redirect_to: destination } );
 		} else {
-			openCheckoutModal( [ plan ], { redirect_to: destination } );
+			openCheckoutModal( [ plan, ...extraProducts ], { redirect_to: destination } );
 		}
```

Comment:
Nitpick (non-blocking): I find this a bit confusing (there is also a small typo). Maybe it's clearer this way?

```js
const products = [
	...( plan ? [ plan ] : [] ),
	...extraProducts,
];

if ( products.length ) {
	openCheckoutModal( products, { redirect_to: destination } );
} else {
	// If no products are provided, we might have added plan to the cart so we just go to the checkout page directly
	// The theme upsell link does not work with siteId and requires a siteSlug.
	// See https://github.com/Automattic/wp-calypso/pull/64899
	window.location.href = `/checkout/${ encodeURIComponent( siteSlug ) }?${ params }`;
}
```

Reply:
Agreed, this is a much clear approach https://github.com/Automattic/wp-calypso/pull/80988/commits/891a4fecf766016a7cb03baf50ab5e4f6bb3b040